### PR TITLE
Add retry handler for the CreateArtifactAsync function

### PR DESCRIPTION
**Pull request by jahsu-MSFT**
https://dev.azure.com/mseng/AzureDevOps/_workitems/edit/1595606/

### CODE REVIEW

File: src/Agent.Plugins/PipelineArtifact/PipelineArtifactServer.cs [hunk 1]
```diff
@@ -313,10 +324,17 @@ public class PipelineArtifactServer
             }
         }
 
-        private CallbackAppTraceSource CreateTracer(AgentTaskPluginExecutionContext context)
-        {
-            var tracer = new CallbackAppTraceSource(str => context.Output(str), System.Diagnostics.SourceLevels.Information);
-            return tracer;
+        public class Tracer {
+            private static CallbackAppTraceSource tracer;
```
Comment (owenhuynMSFT) [inline src/Agent.Plugins/PipelineArtifact/PipelineArtifactServer.cs]: This seems overkill,

I would make Tracer a readonly member variable of PipelineArtifactServer so when you construct PipelineArtifactServer, you will need to pass the AgentContext as part of the constructor which will then fill in the Tracer variable. That way you only have one Tracer in the PipelineArtifactServer which can be recycled across multiple calls.

File: src/Agent.Plugins/PipelineArtifact/BuildServer.cs [hunk 2]
```diff
@@ -17,6 +17,7 @@ public class BuildServer
 
         public BuildServer(VssConnection connection)
         {
+            connection.Settings.SendTimeout = TimeSpan.FromSeconds(50);
```
Comment (johnterickson) [inline src/Agent.Plugins/PipelineArtifact/BuildServer.cs]: Is this higher or lower?

File: src/Agent.Plugins/PipelineArtifact/PipelineArtifactServer.cs [hunk 3]
```diff
@@ -56,17 +63,27 @@ public class PipelineArtifactServer
...
-                                                                        result.ManifestId.ValueString, 
-                                                                        propertiesDictionary, 
-                                                                        cancellationToken);
 
-                context.Output(StringUtil.Loc("AssociateArtifactWithBuild", artifact.Id, pipelineId));
+                BuildArtifact buildArtifact = await AsyncHttpRetryHelper.InvokeAsync(
+                    async () => 
+                    {
+                        return await buildHelper.AssociateArtifactAsync(projectId, 
+                                                                                pipelineId,
```
Comment (johnterickson) [inline src/Agent.Plugins/PipelineArtifact/PipelineArtifactServer.cs]: nit: text alignment

File: src/Agent.Plugins/PipelineArtifact/BuildServer.cs (hunk 2 above)
Comment (jahsu-MSFT) [reply src/Agent.Plugins/PipelineArtifact/BuildServer.cs]: Updating it to 5min/300secs

Comment (johnterickson) [reply src/Agent.Plugins/PipelineArtifact/BuildServer.cs]: Make it higher :) 5 minutes?

File: src/Agent.Plugins/PipelineArtifact/PipelineArtifactPlugin.cs [hunk 4]
```diff
@@ -24,6 +25,8 @@ public async Task RunAsync(AgentTaskPluginExecutionContext context, Cancellation
             // TODO: Translate targetPath from container to host (Ting)
             string targetPath = context.GetInput(ArtifactEventProperties.TargetPath, required: true);
 
+            this.tracer = new CallbackAppTraceSource(str => context.Output(str), System.Diagnostics.SourceLevels.Information);
```
Comment (johnterickson) [inline src/Agent.Plugins/PipelineArtifact/PipelineArtifactPlugin.cs]: does this cause any output on a call without retries?

Comment (jahsu-MSFT) [reply src/Agent.Plugins/PipelineArtifact/PipelineArtifactPlugin.cs]: I don't think so.

**Comment by jtpetty** [comment]
LGTM - will approve once CI is done

File: src/Agent.Plugins/PipelineArtifact/PipelineArtifactServer.cs [hunk 5]
```diff
@@ -66,7 +76,14 @@ public class PipelineArtifactServer
                                                                         propertiesDictionary, 
                                                                         cancellationToken);
 
-                context.Output(StringUtil.Loc("AssociateArtifactWithBuild", artifact.Id, pipelineId));
+                    },
+                    maxRetries: 3,
```
Comment (arunkm) [inline src/Agent.Plugins/PipelineArtifact/PipelineArtifactServer.cs]: A higher number for retires might be better ?

Comment (jahsu-MSFT) [reply src/Agent.Plugins/PipelineArtifact/PipelineArtifactServer.cs]: This is retry on top of retry. I don't think we want to retry to many times.  And the timeout 300 secs should give the server sufficient time to complete the process.

**Pull request merged**